lms/security.py

```diff
@@ -199,6 +212,56 @@ def permits(self, request, _context, permission):
         return _permits(self.identity(request), permission)
 
 
+class EmailPreferencesSecurityPolicy:
+    """The security policy for the email preferences page."""
+
+    def __init__(
+        self,
+        secret: str,
+        domain: str,
+        email_preferences_service: EmailPreferencesService,
+    ):
+        self.cookie = AuthTktCookieHelper(
+            secret=secret,
+            cookie_name="email.preferences",
+            secure=True,
+            timeout=60 * 5,
+            reissue_time=0,
+            max_age=60 * 5,
+            path="/email/preferences",
+            http_only=True,
+            domain=domain,
+        )
+        self.email_preferences_service = email_preferences_service
+
+    def identity(self, request):
+        try:
+            return self.email_preferences_service.h_userid(request.url)
+        except UnrecognisedURLError:
+            pass
+        except InvalidTokenError:
+            return None
+
+        try:
+            return self.cookie.identify(request)["userid"]
```

Comment:
This still has the same problem I reported on the previous draft.

The value returned by `self.cookie.identify(request)["userid"]` is a string, not a `dict`. I suppose it has something to do with how the `AuthTktCookieHelper` serializes/deserializes data, but could not confirm this.

In this branch that does not seem to be a problem, but with the changes I'm trying to apply [here](https://github.com/hypothesis/lms/pull/5860/files#diff-9ebb1a0589d0352360611b3566353720afa6cb2b654494a298342f75c0a87bc6R81), `request.identiy.lti_user` ends up being called, throwing `AttributeError: 'str' object has no attribute 'lti_user'`.

Reply:
Actually, I think we should be wrapping that in `Identity(...)`? 🤔

Reply:
Hmm. I wonder if some wires are getting cross somehow. There's an `lms.security.Identify` class that's used by some of the LMS app's security policies. The new `EmailPreferencesSecurityPolicy` *could* use that `Identity` class and maybe it should, but I don't think it has to.

The email preferences page-loads and form submissions are authenticated by a cookie that's handled between the backend and the browser, I don't think the frontend should need to know anything about that?

So then all the frontend needs is the boolean values for the seven week days. The existing preferences view already passes these values to the template:

https://github.com/hypothesis/lms/blob/fd460ce4712669b3d64dc9a2ac8f8b25c97a3dc7/lms/views/email.py#L74-L84

The template currently uses these template variables to render an HTML form directly. But it could instead render a `js-config` object, something like this in the `preferences.html.jinja2` template:

```html
<script class="js-config">
  days = {
    {% for key, value in preferences.items() %}
    {{ key }}: "{{ value }}"
    {% endfor %}
  }
</script>
```

I probably wouldn't involve the Python `JSConfig` class (`self.context.js_config`) in this. The email preferences page is a standalone thing and doesn't need to be entangled with that.

Reply:
> The new `EmailPreferencesSecurityPolicy` _could_ use that `Identity` class and maybe it should, but I don't think it has to.

I thought there could be some kind of common `SecurityPolicy` contract that was neglected here accidentally, but good from my side if this was intentional.

> I probably wouldn't involve the Python `JSConfig` class (`self.context.js_config`) in this. The email preferences page is a standalone thing and doesn't need to be entangled with that.

Yeah, I thought this could perfectly be the case. Since we want to have a new client app, that involves a new `mode`, and since all modes were being handled there, I thought it made sense to re-use that.

Reply:
Is it possible to use JSConfig in a relatively lightweight way, see:

https://github.com/hypothesis/lms/blob/main/lms/resources/_js_config/__init__.py#L176

Not sure if you'll find any assumptions about being in a LTI launch context there. I reckon is worth a try.

Reply:
I'm afraid the "crash" happens due to this instruction in `JSConfig` initializer https://github.com/hypothesis/lms/blob/main/lms/resources/_js_config/__init__.py#L34, so it's not possible to use this class if `request.identity` is a string.

That's why I made the assumption there was an implicit contract that was not being followed here.

When `request.lti_user` is called, it calls [get_lti_user](https://github.com/hypothesis/lms/blob/main/lms/security.py#L233-L253) from the security module (it gets added to the request [here](https://github.com/hypothesis/lms/blob/main/lms/security.py#L264))

I guess one possible solution would be to move that to a private method so that it only gets invoked by those methods that need it, but that would add a lot of implicit knowledge that's easy to forget in the future, so I tend to agree to better not use `JSConfig` for this use case.

Reply:
>  knowledge that's easy to forget in the future, so I tend to agree to better not use JSConfig for this use case.

Yeah, that's fair enough :+1: